### PR TITLE
Migrate Notebook visualization functionality from Javis to JavisNB

### LINKED ISSUE #1: Migrate Jupyter Utilities from Javis to JavisNB
**Is your feature request related to a problem? Please explain.**

Currently, Javis has Jupyter + Interact utilities inside of the package.
It would be best to move those out to JavisNB.

**Describe the solution you'd like**

Move the lines starting here: https://github.com/JuliaAnimators/Javis.jl/blob/bc65d7a29fcf80e92b08b2455e699d65b822f718/src/javis_viewer.jl#L241

Pertaining to Jupyter to JavisNB!
Make sure to also migrate the tests as well! 

Steps:

- [ ] Migrate Jupyter utilities
- [ ] Create tests
- [ ] Add documentation

**Describe alternatives you've considered**

N/A

**Additional context**

Based on a conversation with @gpucce

### PR DESCRIPTION
**Pull request by gpucce**
## PR Checklist

If you are contributing to `JavisNB.jl`, please make sure you are able to check off each item on this list:

- [ ] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [ ] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [ ] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [ ] Did I properly add `JavisNB` dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [ ] Did I properly add test dependencies to the `test` directory (if applicable)?
- [ ] Did I check relevant tutorials that may be affected by changes in this PR?
- [ ] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**

Closes #1, #2, #3

**How did you address these issues with this PR? What methods did you use?**
I started moving the functionality to render the videos within the notebooks. 

The idea is to use the `embed` function that, if called in place of the `Javis.render` one will replicate the current functionalities of `Javis.render` within a notebook. That means, what used to be

```julia
using Javis
render(...)
```
now becomes

```julia
using Javis, JavisNB
embed(...)
```

At the same time I created a branch in [my fork](https://github.com/gpucce/Javis.jl) called `gp-move-viewers-to-separate-package` where I started changing `Javis.render`  to accomodate for this, in particular `Javis.render` only returns the pathname without calling the viewer functions even if called within a notebook, same goes for the case that `liveview=true`.

### CODE REVIEW

File: Project.toml
Comment (Wikunia) [inline Project.toml]: We want to allow 1.6 as it will be the new LTS

**Review by Wikunia — COMMENTED** [review]
I know it's only a draft sorry for my short review.
Do we need the `embed` of Liveview? I mean either you want to embed it or you want to open it with the liveviewer, right?

**Comment by gpucce** [comment]
> I know it's only a draft sorry for my short review. Do we need the `embed` of Liveview? I mean either you want to embed it or you want to open it with the liveviewer, right?

So you mean I should use a different function smth like `JavisNB.liveview` complementary to `JavisNB.embed` or should I keep the liveview entirely inside `Javis` ? I would like to move it to `JavisNB` so it all works the same

File: Project.toml
Comment (gpucce) [reply Project.toml]: Will fix!

**Comment by Wikunia** [comment]
I'm unsure whether we want an extra repository for Liveviewer as it's not actually a notebook so I wouldn't put it in this repo. It uses Gtk however and can cause problems on a headless machine so I would go for `JavisViewer.jl` 
There I would call the `render` function liveview. 
In `JavisNB` I like the `embed(render` as you suggested.

**Comment by gpucce** [comment]
> I'm unsure whether we want an extra repository for Liveviewer as it's not actually a notebook so I wouldn't put it in this repo. It uses Gtk however and can cause problems on a headless machine so I would go for `JavisViewer.jl` There I would call the `render` function liveview. In `JavisNB` I like the `embed(render` as you suggested.

Perhaps I later understood what you mean, the thing is that right now setting `liveview=true` has different functionalities inside or outside notebooks. Inside a notebook it makes `render` output the gif as a vector of images, in the REPL it opens the liveviewer

**Comment by Wikunia** [comment]
Yeah that's true. I think I would change that behavior as it's not that intuitive for me. Maybe the embed function itself can have a keyword which outputs it as array. Not sure whether that's feasible codewise though. 

Just from the name of functions the best I can think of is the following:
`render` renders the files to gif or MP4 
`embed` shows the result of a render in the notebook
`filmstrip` or something similar shows the array of frames in the notebook. We could have a function `embed` which takes a filmstrip as an argument. 

What do you think @gpucce @TheCedarPrince

**Comment by gpucce** [comment]
I like the naming scheme! 

As you said I also don't have many ideas to move the `liveview/filmstrip` kwarg from `render` to `embed` without making too many changes to `Javis`.  

One way could be to make `embed` take all `render` arguments and then add to it the `filmstrip` argument that is then passed to `render` through `liveview` this way we would also not have to write `embed(render(...))` though it may be a little confusing having two functions that are nearly identical.

**Comment by TheCedarPrince** [comment]
Hey @gpucce !

Haha, I saw you open this issue and am super thrilled!
Thanks for working on this with us!
Didn't want to comment yet until you felt ready to convert this from a draft to a review PR.

I will say that I really like not having to write `embed(render(...))`.
That syntax does feel a bit clunky to me personally so I am very happy with the direction this PR is going.

To @Wikunia 's points, I really like the naming convention!
Keeping `render` and limiting it to just generating files is great.
`filmstrip` is a very interesting idea as I think this would be better suited to actually be housed inside of Javis.jl itself as opposed to `JavisNB`. 
My rationale for why I think it should be in Javis.jl is that I could see this providing a useful functionality for users who may want to do further image processing with a given animation before it either gets rendered or embedded somewhere - it enables a sort of post-processing if you will.

As a result, it makes sense to have `embed` not only ingest a `FilmStrip` type but to also have `render` ingest a `FilmStrip` type.
What I imagine the `FilmStrip` type could look like is something like this:

```julia
struct FilmStrip
    film::Array,
    framerate::Real,
end
```

I like the syntax of `embed` as well being available within notebooks. 


As a final thought, I think it is also necessary at this time to move the  Javis Viewer out of Javis (with its associated dependencies) as I think worrying about how to best support `liveview` may be distracting from what we are trying to solve in this PR. 
I would say it doesn't matter if we break `liveview` in the process of these PRs as I think it best to move Javis Viewer into its own package `JavisViewer.jl` as @Wikunia suggested. 

Just some thoughts off hand!
Great work so far @gpucce !

**Comment by gpucce** [comment]
Hi @TheCedarPrince thanks for the review and no worries about commenting.

Just to make sure, you say you would like the `FilmStrip` struct to be part of Javis instead of JavisNB right?

Also just in case you did not see them there are the `postprocess_frame` and `postprocess_frames_flow` arguments for `render` that might allow to do the postprocessing you mentioned.

For now I will move one with making `embed` a fully capable `render` for notebooks and test it.

**Comment by TheCedarPrince** [comment]
That is correct @gpucce - what do you think @Wikunia ? 
Would `FilmStrip` make a nice addition?

Also, I completely forgot about the `postprocess` functions.

But that sounds like a good plan of action!
Feel free to ask more questions!

**Comment by Wikunia** [comment]
Yes Filmstrip should be part of Javis and having a render function for a filmstrip would be cool to change the frame rate later on as an example or create several film strips using the filmstrip function and combine them to a single video with the render function. That would have made my life easier when creating my YouTube video.

**Comment by Wikunia** [comment]
Which means I wouldn't store the framerate in the filmstrip itself by the way.

**Comment by gpucce** [comment]
Ok nice, though I guess that will go in a different PR, so for now `embed` will be the notebook equivalent of `render` once `Javis` has `FilmStrip` I will add support in `JavisNB`

**Comment by gpucce** [comment]
I think the main things are there as we said, I would like to add a couple tests but it is impossible (I can't manage) to test having a specific branch as a dependency so I can't use the Javis branch where I am doing the changes to work with JavisNB. Thus, for those tests I believe I have to wait until that is merged in Javis.

Let me know what you think @Wikunia @TheCedarPrince

**Comment by Wikunia** [comment]
Thanks a lot. You can locally test it though by calling `dev ../Javis.jl` for example and then checkout the right branch there. Not sure about on GitHub actions though.

**Comment by gpucce** [comment]
The thing is that I would need to do that in both JavisNB and the test environment and if I did that it throws this strange error that says 'can not merge projects' Googling it it looks like it is not possible. Don't know if you ever had the same issue, however I can try in the next days.

**Comment by Wikunia** [comment]
Hmm that's weird thought it would be simple but I can also check during the weekend

File: .github/workflows/ci.yml
Comment (TheCedarPrince) [inline .github/workflows/ci.yml:13]: Out of curiosity, why was this line added for version?

File: src/JavisNB.jl
Comment (TheCedarPrince) [inline src/JavisNB.jl:45]: Why do we create a dispatch on `show` here for `PlutoViewer`?
Is this to enable one to show a buffer within Pluto?

File: src/JavisNB.jl
Comment (TheCedarPrince) [inline src/JavisNB.jl]: I have to say that I am opposed to this syntax for embed as to me it looks odd from a user interface point of view.
An alternative I am wondering about is if it is possible to instead have a function called `notebook_render` that does not have the syntax for `liveview` but is exclusive to `JavisNB` and always assumes that whenever this function is called, to embed it in a given notebook.
This gets away from the splatting required (I think) and leaves one with the nice syntax of `notebook_render(myVideo)` as really, the function is just a special wrapper around the `Javis.render` command.
What do you think?

I am also a bit unclear as to why splatting is needed - is it to pass the values of `Object` and such?
I may just be missing something here because all it seems to me is to pass a `String` into `embed` at the moment.

**Review by TheCedarPrince — CHANGES_REQUESTED** [review]
Hey @gpucce 

Thank you so much for taking the time to work on this PR!
I had some clarification questions, but in general I really like where this PR is going, but I am still opposed to having the syntax of `embed(render(myvideo)...)` - the splatting just does not feel like well suited for the user interface of Javis.
However, I am happy to keep working with you on this!
I may just be missing something.
Either way, great work and I am looking forward to seeing how we can develop this PR!

~ tcp :deciduous_tree:

File: src/JavisNB.jl
Comment (gpucce) [reply src/JavisNB.jl]: Hi @TheCedarPrince, so the splatting is gone, it was left in the docstring because I was thinking of making a different PR to add all the docs. There was however a small issue with livewview but it is now gone and the docstring for embed now look also better.

Let me know what you think

File: src/JavisNB.jl
Comment (gpucce) [reply src/JavisNB.jl:45]: Yes I believe it is because of how Pluto deals with `show` however this I took from Javis without changing much.

File: .github/workflows/ci.yml
Comment (gpucce) [reply .github/workflows/ci.yml:13]: Looking at when it was added in Javis, I believe it checks for 1.7, so now we should have tests on 1.6 (LTS) and 1.7.smth latest with these lines. Though I might be wrong here.

Comment (TheCedarPrince) [reply .github/workflows/ci.yml:13]: No problem!
Yea, it checks for the latest stable release from Julia which is 1.7 so that works.
We will just need to bump in the future (which is what we do manually for Javis anyways).
Looks good!

File: src/JavisNB.jl
Comment (TheCedarPrince) [reply src/JavisNB.jl]: Ah this makes a lot of sense because I was having issues when trying to run this from within a Pluto notebook!
Thanks!

**Comment by TheCedarPrince** [comment]
Hey @gpucce ,

I just tried this out in a Pluto notebook and am getting this strange error:

image

Any idea as to why this is happening?

EDIT: I can confirm, this error appears to only occur within Pluto notebooks as it works fine within Jupyter notebooks:

image

**Comment by gpucce** [comment]
Hi @TheCedarPrince yeah the error is there because you should use Javis as it is in the PR to move the notebook functionality to JavisNB, 

https://github.com/JuliaAnimators/Javis.jl/pull/451

Sorry I forgot to mention this.

I mean I think that is why, if you are already doing it then I have to check!

File: src/JavisNB.jl
Comment (TheCedarPrince) [inline src/JavisNB.jl]: You can probably delete this as it is not accurate any more.
Furthermore, maybe just say 

```julia
"""
    _pluto_viewer(video::Video, frames::Int, actions::Vector)
Creates an interactive viewer in a Pluto Notebook by storing all the frames in-memory

# Returns

`arr`: List of Javis frames
```

File: src/JavisNB.jl
Comment (TheCedarPrince) [inline src/JavisNB.jl]: Small grammar fixes:

"This is the core function of `JavisNB` it allows to show the output of `render` 
from [`Javis`](https://juliaanimators.github.io/Javis.jl/stable/) directly within 
a notebook (IJulia or Pluto)." 

Fixed to:

```
"The core function of `JavisNB`.
It allows one to show the output of `Javis.render` 
from [`Javis`](https://juliaanimators.github.io/Javis.jl/stable/) directly within a notebook.
Notebooks supported are: 

- IJulia
- Pluto
```

Comment (TheCedarPrince) [inline src/JavisNB.jl]: Could you move this docstring to above the corresponding dispatch on line 154?

File: src/JavisNB.jl
Comment (TheCedarPrince) [inline src/JavisNB.jl]: Suggested docstring fix:

```julia
"""
    embed(pathname::String)

Shows a gif within a notebook by calling the path to the gif.

# Arguments

- `pathname::String`: the path to a gif to render in the notebook
"""
```

**Review by TheCedarPrince — CHANGES_REQUESTED** [review]
Hey @gpucce !
I was able to get it working - I can't believe I didn't think to check out that branch but everything works now.
Just some doc changes and I think this is about ready in my opinion!

**Comment by gpucce** [comment]
Hi @TheCedarPrince I did the changes you suggested and went on and added also the Documenter support and the proper documentation let me know what you think and if it works for you!

File: Project.toml
Comment (Wikunia) [inline Project.toml]: I would use "1.6" here which should default to `1.6-2.0` without including `2.0`

Comment (Wikunia) [reply Project.toml]: We also need to specify compat for `Interact` and `Javis`

**Review by Wikunia — COMMENTED** [review]
Looks awesome thanks a lot. Haven't tried it myself yet but gifs look great and see that it worked well for @TheCedarPrince .
Let me know if I should check it out as well more deeply otherwise I leave it to you two.

File: docs/src/index.md
Comment (TheCedarPrince) [inline docs/src/index.md]: Just some minor grammar suggestions/corrections:

```
`JavisNB` is a small package written for use [Javis.jl](https://juliaanimators.github.io/Javis.jl/stable/) within Jupyter and Pluto notebooks.

It is very straightforward to use.
First write all your code for an animation just like you would with `Javis` and then replace `Javis.render` with `JavisNB.embed`.
Alongside sending the rendered animation to a file, the gif will be shown in the notebook.

The `liveview` argument that in `Javis.render` would activate the Javis Live Viewer tool is available in `JavisNB.embed` as well but works differently.
In `Pluto` it returns an array of the frames composing a gif.
In `Jupyter` it creates an interactive view of the animation and lets you scroll through the frames of the gif.
```

File: docs/src/index.md
Comment (TheCedarPrince) [inline docs/src/index.md]: "Cells' code to copy and paste it:"

to

"Copy and paste each of the code blocks into different cells"

File: docs/src/index.md
Comment (TheCedarPrince) [inline docs/src/index.md]: "Cells' code to copy and paste it:"

to

"Copy and paste each of the code blocks into different cells"

File: docs/src/index.md
Comment (TheCedarPrince) [inline docs/src/index.md]: "Cells' code to copy and paste it:"

to

"Copy and paste each of the code blocks into different cells"

File: docs/src/index.md
Comment (TheCedarPrince) [inline docs/src/index.md]: "Cells' code to copy and paste it:"

to

"Copy and paste each of the code blocks into different cells"

File: src/JavisNB.jl
Comment (TheCedarPrince) [inline src/JavisNB.jl]: Add also:

```
# Returns

- `pathname::String`: path of the gif to render in the notebook
```

File: src/JavisNB.jl
Comment (TheCedarPrince) [inline src/JavisNB.jl]: "To use it just replace `render` with `embed`, an example within `Pluto`:"

to

"To use `embed` just replace `Javis.render` with `embed`.
Here is an example one can use within a `Pluto` notebook:"

**Review by TheCedarPrince — CHANGES_REQUESTED** [review]
Just a few grammar changes!
In the future, if it is alright @gpucce , I can do the grammar fixes for you on your branch.
Once these grammar fixes are made, you have my full approval to merge!
Looks outstanding! :D

**Comment by gpucce** [comment]
Hi @TheCedarPrince grammar changes should be done. I can't merge it myself so waiting for you on that.

For the grammar fixes in general if you tell me I use it to refresh english a bit but if you want to make them and go on it is totally fine!!

**Comment by gpucce** [comment]
Hi @TheCedarPrince 

I changed the docs adding more info, for sure there will be more grammar checking to do sorry for my english :)

But I think it works better now I took some more of the things that were there in the Javis workflows documentation and added them here.

Oh and I also uncommented the CI docs workflow.

**Review by TheCedarPrince — APPROVED** [review]
Looks Good To Me!!!
Let's merge it!
And also let's get you merge access :sweat_smile:

**Pull request merged**
